File: Xamarin/Xamarin.Native/Xamarin.Native.Droid/Activities/LocationTestActivity.cs
using Android.App;
using Android.Content;
using Android.OS;
using Android.Runtime;
using Android.Views;
using Android.Widget;
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Text;
using Xamarin.iOS;

namespace Xamarin.Native.Droid.Activities
{
    [Activity(Label = "LocationActivity")]
    public class LocationTestActivity : Activity
    {
        Stopwatch stopwatch;
        LocationTestService locationService;
        Button startButton;
        TextView TimeLabel;
        TextView PositionLabel;
        ProgressDialog progressDialog;

        protected override void OnCreate(Bundle savedInstanceState)
        {
            base.OnCreate(savedInstanceState);
            SetContentView(Resource.Layout.LocationTest);
            Title = GetString(Resource.String.menuLocation);

            locationService = new LocationTestService();
            locationService.LocationChanged += LocationService_LocationChanged;

            TimeLabel = FindViewById<TextView>(Resource.Id.TimeLabel);
            PositionLabel = FindViewById<TextView>(Resource.Id.PositionLabel);
            startButton = FindViewById<Button>(Resource.Id.StartButton);
            startButton.Click += StartPositioning;
        }

        private void StartPositioning(object sender, EventArgs e)
        {
            stopwatch = new Stopwatch();
            progressDialog = ProgressDialog.Show(this, "Wyznaczanie pozycji...", "");

            stopwatch.Start();
            locationService.GetLocation();
        }

        private void LocationService_LocationChanged(double latitude, double longitude)
        {
            stopwatch.Stop();
            RunOnUiThread(() =>
            {
                PositionLabel.Text = string.Format("D³ugoœæ: {0}\nSzerokoœæ: {1}", Math.Round(longitude, 4), Math.Round(latitude, 4));
                TimeLabel.Text = stopwatch.GetDurationInSeconds();
                progressDialog.Dismiss();
            });
        }
    }
}

File: Native.Android/app/src/main/java/com/pgrzmil/nativedroid/LocationTestActivity.java
package com.pgrzmil.nativedroid;

import android.app.ProgressDialog;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.view.View;
import android.widget.*;
import android.app.Activity;

import com.pgrzmil.services.*;

import java.math.RoundingMode;
import java.text.DecimalFormat;

public class LocationTestActivity extends Activity implements LocationTestListener {
    Stopwatch stopwatch;
    LocationTestService locationService;
    Button startButton;
    TextView timeLabel;
    TextView positionLabel;
    ProgressDialog progressDialog;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_location);
        setTitle(R.string.menuLocation);

        timeLabel = (TextView) findViewById(R.id.timeLabel);
        positionLabel = (TextView) findViewById(R.id.positionLabel);
        startButton = (Button) findViewById(R.id.startButton);

        stopwatch = new Stopwatch();

        locationService = new LocationTestService(this);
        locationService.addListener(this);
    }

    public void startPositioning(View view) {
        progressDialog = ProgressDialog.show(this, "Wyznaczanie pozycji...", "");
        stopwatch.start();

        locationService.getLocation();
    }

    @Override
    public void locationChanged(final double latitude, final double longitude) {
        stopwatch.stop();
        locationService.stop();

        runOnUiThread(new Runnable() {
            public void run() {
                DecimalFormat formatter = new DecimalFormat("#.####");
                formatter.setRoundingMode(RoundingMode.CEILING);
                positionLabel.setText(String.format("Długość: %s\nSzerokość: %s", formatter.format(longitude), formatter.format(latitude)));
                timeLabel.setText(String.format("%s\n%s", stopwatch.getDurationInSeconds(), stopwatch.getDurationInMilliseconds()));
                progressDialog.dismiss();
            }
        });
    }
}
